src/array/boolean/mutable.rs

```diff
@@ -240,33 +297,53 @@ impl MutableBooleanArray {
 /// # Safety
 /// The caller must ensure that `iterator` is `TrustedLen`.
 #[inline]
-pub(crate) unsafe fn trusted_len_unzip<I, P>(iterator: I) -> (MutableBitmap, MutableBitmap)
+pub(crate) unsafe fn trusted_len_unzip<I, P>(iterator: I) -> (Option<MutableBitmap>, MutableBitmap)
 where
     P: std::borrow::Borrow<bool>,
     I: Iterator<Item = Option<P>>,
+{
+    let mut validity = MutableBitmap::new();
+    let mut values = MutableBitmap::new();
+
+    extend_trusted_len_unzip(iterator, &mut validity, &mut values);
+
+    let validity = if validity.null_count() > 0 {
+        Some(validity)
+    } else {
+        None
+    };
+
+    (validity, values)
+}
+
+/// Extends validity [`MutableBitmap`] and values [`MutableBitmap`] from an iterator of `Option`.
+/// # Safety
+/// The caller must ensure that `iterator` is `TrustedLen`.
+#[inline]
+pub(crate) unsafe fn extend_trusted_len_unzip<I, P>(
+    iterator: I,
+    validity: &mut MutableBitmap,
+    values: &mut MutableBitmap,
+) where
+    P: std::borrow::Borrow<bool>,
+    I: Iterator<Item = Option<P>>,
 {
     let (_, upper) = iterator.size_hint();
-    let len = upper.expect("trusted_len_unzip requires an upper limit");
+    let additional = upper.expect("extend_trusted_len_unzip requires an upper limit");
 
-    let mut validity = MutableBitmap::with_capacity(len);
-    let mut values = MutableBitmap::with_capacity(len);
+    validity.reserve(additional);
+    values.reserve(additional);
 
     for item in iterator {
         let item = if let Some(item) = item {
-            validity.push(true);
+            validity.push_unchecked(true);
             *item.borrow()
         } else {
-            validity.push(false);
-            false
+            validity.push_unchecked(false);
+            bool::default()
         };
-        values.push(item);
+        values.push_unchecked(item);
     }
-    assert_eq!(
```

Comment:
Could we keep the `debug_assert_eq!` here?

Reply:
@jorgecarleitao I have added `debug_assert_eq`, but I had to create a variable to store the length value before expansion and use it later in the assert statement